File: src/subc/parse/decl.go
package parse

import (
	"fmt"

	"subc/ast"
	"subc/scan"
)

/*
 * top :=
 *	  ENUM enumdecl
 *	| decl
 *	| primtype decl
 *	| storclass decl
 *	| storclass primtype decl
 *
 * storclass :=
 *	  EXTERN
 *	| STATIC
 *	| VOLATILE
 */

func (p *parser) top() (decls []ast.Decl) {
	var storage *scan.Token
	switch tok := p.peek(); tok.Type {
	case scan.Extern, scan.Static, scan.Volatile:
		storage = &tok
		p.next()
	}

	switch tok := p.peek(); tok.Type {
	case scan.Enum:
		d := p.enumDecl(true)
		d.Storage = storage
		decls = append(decls, d)
	case scan.Struct, scan.Union:
		decls = p.structDecl(storage)
	case scan.Char, scan.Int, scan.Void:
		p.next()
		decls = p.decl(storage, &ast.BasicType{Type: tok})
	case scan.Ident:
		decls = p.decl(storage, nil)
	default:
		p.errorf(tok.Pos, "expected type specifier, got %v", tok.Type)
		span := p.synch(tok.Pos, scan.Semi)
		badDecl := &ast.BadDecl{span.Start, span.End}
		decls = append(decls, badDecl)
	}

	return
}

/*
 * enumdecl := { enumlist } ;
 *
 * enumlist :=
 *	  enumerator
 *	| enumerator , enumlist
 *
 * enumerator :=
 *	  IDENT
 *	| IDENT = constexpr
 */

func (p *parser) enumDecl(global bool) *ast.EnumDecl {
	d := &ast.EnumDecl{}
	d.Enum = p.next()
	if tok := p.peek(); tok.Type == scan.Ident {
		d.Name = &ast.Ident{tok.Pos, tok.Text}
		p.next()
	}
	d.Lbrace = p.expect(scan.Lbrace)
	for {
		if tok := p.peek(); tok.Type == scan.Rbrace {
			break
		}

		e := &ast.ConstDecl{}
		name := p.expect(scan.Ident)
		e.Name = &ast.Ident{name.Pos, name.Text}
		if tok := p.peek(); tok.Type == scan.Assign {
			p.next()
			e.X = p.constExpr()
		}

		d.List = append(d.List, e)

		if tok := p.peek(); tok.Type != scan.Comma {
			break
		}

		if p.eofCheck() {
			return d
		}
		p.next()
	}
	d.Rbrace = p.expect(scan.Rbrace)
	p.expect(scan.Semi)
	return d
}

/*
 * initlist :=
 *	  { const_list }
 *	| STRLIT
 *
 * const_list :=
 *	  constexpr
 *	| constexpr , const_list
 */

func (p *parser) initList() ast.Expr {
	s := &ast.StringLit{}
	for {
		tok := p.peek()
		if tok.Type != scan.String {
			break
		}
		s.Lits = append(s.Lits, &ast.BasicLit{tok})
		p.next()
	}
	if len(s.Lits) > 0 {
		return s
	}

	e := &ast.CompositeLit{}
	e.Lbrace = p.expect(scan.Lbrace)
	for {
		if tok := p.peek(); tok.Type == scan.Rbrace {
			e.Rbrace = tok
			break
		}
		e.Elts = append(e.Elts, p.constExpr())

		if tok := p.peek(); tok.Type == scan.Comma {
			p.next()
		} else {
			break
		}

		if p.eofCheck() {
			return e
		}
	}
	p.expect(scan.Rbrace)
	return e
}

/*
 * structdecl :=
 *	  STRUCT identifier { member_list } ;
 *
 * member_list :=
 *	  primtype mdecl_list ;
 *	| primtype mdecl_list ; member_list
 *
 * mdecl_list :=
 *	  declarator
 *	| declatator , mdecl_list
 */

func (p *parser) structDecl(storage *scan.Token) (decls []ast.Decl) {
	typ := p.next()
	name := p.expect(scan.Ident)
	ident := &ast.Ident{name.Pos, name.Text}
	if tok := p.peek(); tok.Type != scan.Lbrace {
		prim := &ast.RecordType{Record: typ, Name: ident}
		decls = p.decl(storage, prim)
		return
	}

	d := &ast.RecordDecl{}
	d.Record = typ
	d.Name = &ast.Ident{name.Pos, name.Text}
	d.Lbrace = p.next()

	for {
		tok := p.peek()
		if !isType(tok.Type) {
			break
		}

		prim := p.primType(tok)

		for {
			if p.eofCheck() {
				decls = append(decls, d)
				return
			}

			v := p.declarator(true, storage, prim).(*ast.VarDecl)
			f := &ast.FieldDecl{Type: prim}
			f.Name = v.Name
			f.Type = v.Type
			d.Fields = append(d.Fields, f)

			if tok := p.peek(); tok.Type != scan.Comma {
				break
			}
			p.next()
		}
		p.expect(scan.Semi)
	}
	d.Rbrace = p.expect(scan.Rbrace)
	p.expect(scan.Semi)
	decls = append(decls, d)
	return
}

/*
 * decl :=
 *	  declarator { localdecls stmt_list }
 *	| decl_list ;
 *
 * decl_list :=
 *	  declarator
 *	| declarator , decl_list
 */

func (p *parser) decl(storage *scan.Token, prim ast.Decl) (decls []ast.Decl) {
	for {
		d := p.declarator(false, storage, prim)
		decls = append(decls, d)
		switch d := d.(type) {
		case *ast.FuncDecl:
			if tok := p.peek(); tok.Type == scan.Lbrace {
				p.curFn = d
				p.next()
				d.Decls = p.localDecls()
				d.Body = p.compound(&tok)
			} else {
				p.expect(scan.Semi)
			}
			p.curFn = nil
			return
		}

		if tok := p.peek(); tok.Type == scan.Comma {
			p.next()
		} else {
			break
		}
	}

	p.expect(scan.Semi)
	return
}

/*
 * declarator :=
 *	  IDENT
 *	| * IDENT
 *	| * * IDENT
 *	| * IDENT [ constexpr ]
 *	| IDENT [ constexpr ]
 *	| IDENT = constexpr
 *	| IDENT [ ] = initlist
 *	| IDENT pmtrdecl
 *	| IDENT [ ]
 *	| * IDENT [ ]
 *	| ( * IDENT ) ( )
 */

func (p *parser) declarator(pmtr bool, storage *scan.Token, prim ast.Decl) ast.Decl {
	var f *ast.FuncType
	var a *ast.ArrayType
	var s *ast.StarExpr
	v := &ast.VarDecl{Storage: storage, Type: copyPrim(prim)}

	switch tok := p.peek(); tok.Type {
	case scan.Mul:
		star := &ast.StarExpr{Star: tok}
		s = star
		p.next()
		switch tok := p.peek(); tok.Type {
		case scan.Mul:
			s = &ast.StarExpr{Star: tok}
			star.X = s
			p.next()
		}

		setType(v.Type, star)

	case scan.Lparen:
		f = &ast.FuncType{Result: copyPrim(prim)}
		f.Lparen[0] = tok
		p.next()
		f.Star = p.expect(scan.Mul)
	}

	if tok := p.peek(); !pmtr || tok.Type == scan.Ident {
		name := p.expect(scan.Ident)
		v.Name = &ast.Ident{name.Pos, name.Text}
	}

	if f != nil {
		v.Type = f
		f.Rparen[0] = p.expect(scan.Rparen)
		f.Lparen[1] = p.expect(scan.Lparen)
		f.Rparen[1] = p.expect(scan.Rparen)
	}

	d := ast.Decl(v)
	switch tok := p.peek(); {
	case !pmtr && tok.Type == scan.Assign:
		p.next()
		v.Value = p.constExpr()

	case !pmtr && tok.Type == scan.Lparen:
		fd := &ast.FuncDecl{}
		fd.Storage = storage
		fd.Result = v.Type
		fd.Name = v.Name
		fd.Lparen = p.next()
		fd.Params = p.pmtrDecls()
		fd.Rparen = p.expect(scan.Rparen)
		d = fd

	case tok.Type == scan.Lbrack:
		a = &ast.ArrayType{}
		a.Lbrack = tok
		p.next()
		if tok := p.peek(); tok.Type == scan.Rbrack {
			a.Rbrack = tok
			p.next()
			if tok := p.peek(); tok.Type == scan.Assign {
				p.next()
				v.Value = p.initList()
			}
		} else {
			a.Len = p.constExpr()
			a.Rbrack = p.expect(scan.Rbrack)
		}

		if s != nil {
			s.X = a
		} else {
			setType(v.Type, a)
		}
	}

	return d
}

/*
 * pmtrdecl :=
 *	  ( )
 *	| ( pmtrlist )
 *	| ( pmtrlist , ... )
 *
 * pmtrlist :=
 *	  primtype declarator
 *	| primtype declarator , pmtrlist
 */

func (p *parser) pmtrDecls() (fields []*ast.FieldDecl) {
	if tok := p.peek(); tok.Type == scan.Rparen {
		return nil
	}

	var prim ast.Decl
	nargs := 0

loop:
	for {
		switch tok := p.peek(); {
		case nargs > 0 && tok.Type == scan.Ellipsis:
			fields = append(fields, &ast.FieldDecl{Type: tok})
			p.next()
			break loop

		case tok.Type == scan.Ident:

		default:
			if !isType(tok.Type) {
				p.errorf(tok.Pos, "type specifier expected")
				p.synch(tok.Pos, scan.Rparen)
				break loop
			}

			xtok := tok
			prim = p.primType(tok)
			if tok := p.peek(); tok.Type == scan.Rparen && xtok.Type == scan.Void && nargs == 0 {
				fields = append(fields, &ast.FieldDecl{Type: prim})
				break loop
			}
		}

		v := p.declarator(true, nil, prim)
		switch v := v.(type) {
		case *ast.VarDecl:
			field := &ast.FieldDecl{Type: v.Type, Name: v.Name}
			fields = append(fields, field)

		default:
			panic(fmt.Sprintf("unimplemented parameter type: %T", v))
		}

		nargs++
		if tok := p.peek(); tok.Type == scan.Comma {
			p.next()
		} else {
			break
		}
	}

	return
}

/*
 * localdecls :=
 *        ldecl
 *      | ldecl localdecls
 *
 * ldecl :=
 *	  primtype ldecl_list ;
 *	| lclass primtype ldecl_list ;
 *	| lclass ldecl_list ;
 *	| enum_decl
 *	| struct_decl
 *
 * lclass :=
 *	| AUTO
 *	| EXTERN
 *	| REGISTER
 *	| STATIC
 *	| VOLATILE
 *
 * ldecl_list :=
 *	  declarator
 *	| declarator , ldecl_list
 */

func (p *parser) localDecls() []ast.Decl {
	var d []ast.Decl

	for {
		tok := p.peek()
		if !isLocalType(tok.Type) {
			break
		}

		if tok.Type == scan.Enum {
			d = append(d, p.enumDecl(false))
			continue
		}

		var storage *scan.Token
		var prim ast.Decl
		if isQualifier(tok.Type) {
			storage = &tok
			p.next()
			if tok := p.peek(); isType(tok.Type) {
				prim = p.primType(tok)
			}
		} else {
			prim = p.primType(tok)
		}

		for {
			if p.eofCheck() {
				return d
			}

			n := p.declarator(false, storage, prim)
			d = append(d, n)

			if tok := p.peek(); tok.Type == scan.Comma {
				p.next()
			} else {
				break
			}
		}
		p.expect(scan.Semi)
	}

	return d
}

// primType creates a basic type or a record type from token
func (p *parser) primType(tok scan.Token) ast.Expr {
	switch tok.Type {
	case scan.Char, scan.Int, scan.Void:
		p.next()
		return &ast.BasicType{Type: tok}

	case scan.Union, scan.Struct:
		p.next()
		name := p.expect(scan.Ident)
		ident := &ast.Ident{name.Pos, name.Text}
		return &ast.RecordType{Record: tok, Name: ident}

	default:
		panic(fmt.Sprintf("unimplemented primary type %v", tok))
	}
}

// setType sets y to be an element that x points to, x needs to be a basic type
// or a record type.
func setType(x, y ast.Expr) {
	switch x := x.(type) {
	case *ast.BasicType:
		x.X = y
	case *ast.RecordType:
		x.X = y
	case nil:
	default:
		panic(fmt.Sprintf("unimplemented type: %T", x))
	}
}

// copyPrim makes a copy of prim.
func copyPrim(prim ast.Expr) ast.Expr {
	var d ast.Expr
	switch x := prim.(type) {
	case *ast.BasicType:
		y := &ast.BasicType{}
		*y = *x
		d = y
	case *ast.RecordType:
		y := &ast.RecordType{}
		*y = *x
		d = y
	case nil:
	default:
		panic(fmt.Sprintf("unimplemented new primary type: %v", x))
	}
	return d
}

File: test/decl.c
/*
 *	NMH's Simple C Compiler, 2011--2014
 *	Declaration parser
 */

#include "defs.h"
#include "data.h"
#include "decl.h"

static int declarator(int arg, int scls, char *name, int *pprim, int *psize,
			int *pval, int *pinit);

/*
 * enumdecl := { enumlist } ;
 *
 * enumlist :=
 *	  enumerator
 *	| enumerator , enumlist
 *
 * enumerator :=
 *	  IDENT
 *	| IDENT = constexpr
 */

static void enumdecl(int glob) {
	int	v = 0;
	char	name[NAMELEN+1];

	Token = scan();
	if (IDENT == Token)
		Token = scan();
	lbrace();
	while (RBRACE != Token) {
		copyname(name, Text);
		ident();
		if (ASSIGN == Token) {
			Token = scan();
			v = constexpr();
		}
		if (glob)
			addglob(name, PINT, TCONSTANT, 0, 0, v++, NULL, 0);
		else
			addloc(name, PINT, TCONSTANT, 0, 0, v++, 0);
		if (Token != COMMA)
			break;
		Token = scan();
		if (eofcheck()) return;
	}
	rbrace();
	semi();
}

/*
 * initlist :=
 *	  { const_list }
 *	| STRLIT
 *
 * const_list :=
 *	  constexpr
 *	| constexpr , const_list
 */

static int initlist(char *name, int prim) {
	int	n = 0, v;
	char	buf[30];

	gendata();
	genname(name);
	if (STRLIT == Token) {
		if (PCHAR != prim)
			error("initializer type mismatch: %s", name);
		gendefs(Text, Value);
		gendefb(0);
		genalign(Value-1);
		Token = scan();
		return Value-1;
	}
	lbrace();
	while (Token != RBRACE) {
		v = constexpr();
		if (PCHAR == prim) {
			if (v < 0 || v > 255) {
				sprintf(buf, "%d", v);
				error("initializer out of range: %s", buf);
			}
			gendefb(v);
		}
		else {
			gendefw(v);
		}
		n++;
		if (COMMA == Token)
			Token = scan();
		else
			break;
		if (eofcheck()) return 0;
	}
	if (PCHAR == prim) genalign(n);
	Token = scan();
	if (!n) error("too few initializers", NULL);
	return n;
}

int primtype(int t, char *s) {
	int	p, y;
	char	sname[NAMELEN+1];

	p = t == CHAR? PCHAR:
		t == INT? PINT:
		t == STRUCT? PSTRUCT:
		t == UNION? PUNION:
		PVOID;
	if (PUNION == p || PSTRUCT == p) {
		if (!s) {
			Token = scan();
			copyname(sname, Text);
			s = sname;
			if (IDENT != Token) {
				error("struct/union name expected: %s", Text);
				return p;
			}
		}
		if ((y = findstruct(s)) == 0 || Prims[y] != p)
			error("no such struct/union: %s", s);
		p |= y;
	}
	return p;
}

/*
 * pmtrdecl :=
 *	  ( )
 *	| ( pmtrlist )
 *	| ( pmtrlist , ... )
 *
 * pmtrlist :=
 *	  primtype declarator
 *	| primtype declarator , pmtrlist
 */

static int pmtrdecls(void) {
	char	name[NAMELEN+1];
	int	prim, type, size, na, addr;
	int	dummy;

	if (RPAREN == Token)
		return 0;
	na = 0;
	addr = 2*BPW;
	for (;;) {
		if (na > 0 && ELLIPSIS == Token) {
			Token = scan();
			na = -(na + 1);
			break;
		}
		else if (IDENT == Token) {
			prim = PINT;
		}
		else {
			if (	Token != CHAR && Token != INT &&
				Token != VOID &&
				Token != STRUCT && Token != UNION
			) {
				error("type specifier expected at: %s", Text);
				Token = synch(RPAREN);
				return na;
			}
			name[0] = 0;
			prim = primtype(Token, NULL);
			Token = scan();
			if (RPAREN == Token && prim == PVOID && !na)
				return 0;
		}
		size = 1;
		type = declarator(1, CAUTO, name, &prim, &size, &dummy,
				&dummy);
		addloc(name, prim, type, CAUTO, size, addr, 0);
		addr += BPW;
		na++;
		if (COMMA == Token)
			Token = scan();
		else
			break;
	}
	return na;
}

int pointerto(int prim) {
	int	y;

	if (CHARPP == prim || INTPP == prim || VOIDPP == prim ||
	    FUNPTR == prim ||
	    (prim & STCMASK) == STCPP || (prim & STCMASK) == UNIPP
	)
		error("too many levels of indirection", NULL);
	y = prim & ~STCMASK;
	switch (prim & STCMASK) {
	case PSTRUCT:	return STCPTR | y;
	case STCPTR:	return STCPP | y;
	case PUNION:	return UNIPTR | y;
	case UNIPTR:	return UNIPP | y;
	}
	return PINT == prim? INTPTR:
		PCHAR == prim? CHARPTR:
		PVOID == prim? VOIDPTR:
		INTPTR == prim? INTPP:
		CHARPTR == prim? CHARPP: VOIDPP;
}

/*
 * declarator :=
 *	  IDENT
 *	| * IDENT
 *	| * * IDENT
 *	| * IDENT [ constexpr ]
 *	| IDENT [ constexpr ]
 *	| IDENT = constexpr
 *	| IDENT [ ] = initlist
 *	| IDENT pmtrdecl
 *	| IDENT [ ]
 *	| * IDENT [ ]
 *	| ( * IDENT ) ( )
 */

static int declarator(int pmtr, int scls, char *name, int *pprim, int *psize,
			int *pval, int *pinit)
{
	int	type = TVARIABLE;
	int	ptrptr = 0;

	if (STAR == Token) {
		Token = scan();
		*pprim = pointerto(*pprim);
		if (STAR == Token) {
			Token = scan();
			*pprim = pointerto(*pprim);
			ptrptr = 1;
		}
	}
	else if (LPAREN == Token) {
		if (*pprim != PINT)
			error("function pointers are limited to type 'int'",
				NULL);
		Token = scan();
		*pprim = FUNPTR;
		match(STAR, "(*name)()");
	}
	if (IDENT != Token) {
		error("missing identifier at: %s", Text);
		name[0] = 0;
	}
	else {
		copyname(name, Text);
		Token = scan();
	}
	if (FUNPTR == *pprim) {
		rparen();
		lparen();
		rparen();
	}
	if (!pmtr && ASSIGN == Token) {
		Token = scan();
		*pval = constexpr();
		if (PCHAR == *pprim)
			*pval &= 0xff;
		if (*pval && !inttype(*pprim))
			error("non-zero pointer initialization", NULL);
		*pinit = 1;
	}
	else if (!pmtr && LPAREN == Token) {
		Token = scan();
		*psize = pmtrdecls();
		rparen();
		return TFUNCTION;
	}
	else if (LBRACK == Token) {
		if (ptrptr)
			error("too many levels of indirection: %s", name);
		Token = scan();
		if (RBRACK == Token) {
			Token = scan();
			if (pmtr) {
				*pprim = pointerto(*pprim);
			}
			else {
				type = TARRAY;
				*psize = 1;
				if (ASSIGN == Token) {
					Token = scan();
					if (!inttype(*pprim))
						error("initialization of"
							" pointer array not"
							" supported",
							NULL);
					*psize = initlist(name, *pprim);
					if (CAUTO == scls)
						error("initialization of"
							" local arrays"
							" not supported: %s",
							name);
					*pinit = 1;
				}
				else if (CEXTERN != scls) {
					error("automatically-sized array"
						" lacking initialization: %s",
						name);
				}
			}
		}
		else {
			*psize = constexpr();
			if (*psize < 0) {
				error("invalid array size", NULL);
				*psize = 0;
			}
			type = TARRAY;
			rbrack();
		}
	}
	if (PVOID == *pprim)
		error("'void' is not a valid type: %s", name);
	return type;
}

/*
 * localdecls :=
 *        ldecl
 *      | ldecl localdecls
 *
 * ldecl :=
 *	  primtype ldecl_list ;
 *	| lclass primtype ldecl_list ;
 *	| lclass ldecl_list ;
 *	| enum_decl
 *	| struct_decl
 *
 * lclass :=
 *	| AUTO
 *	| EXTERN
 *	| REGISTER
 *	| STATIC
 *	| VOLATILE
 *
 * ldecl_list :=
 *	  declarator
 *	| declarator , ldecl_list
 */

static int localdecls(void) {
	char	name[NAMELEN+1];
	int	prim, type, size, addr = 0, val, ini;
	int	stat, extn;
	int	pbase, rsize;

	Nli = 0;
	while ( AUTO == Token || EXTERN == Token || REGISTER == Token ||
		STATIC == Token || VOLATILE == Token ||
		INT == Token || CHAR == Token || VOID == Token ||
		ENUM == Token ||
		STRUCT == Token || UNION == Token
	) {
		if (ENUM == Token) {
			enumdecl(0);
			continue;
		}
		extn = stat = 0;
		if (AUTO == Token || REGISTER == Token || STATIC == Token ||
			VOLATILE == Token || EXTERN == Token
		) {
			stat = STATIC == Token;
			extn = EXTERN == Token;
			Token = scan();
			if (	INT == Token || CHAR == Token ||
				VOID == Token ||
				STRUCT == Token || UNION == Token
			) {
				prim = primtype(Token, NULL);
				Token = scan();
			}
			else
				prim = PINT;
		}
		else {
			prim = primtype(Token, NULL);
			Token = scan();
		}
		pbase = prim;
		for (;;) {
			prim = pbase;
			if (eofcheck()) return 0;
			size = 1;
			ini = val = 0;
			type = declarator(0, CAUTO, name, &prim, &size,
					&val, &ini);
			rsize = objsize(prim, type, size);
			rsize = (rsize + INTSIZE-1) / INTSIZE * INTSIZE;
			if (stat) {
				addloc(name, prim, type, CLSTATC, size,
					label(), val);
			}
			else if (extn) {
				addloc(name, prim, type, CEXTERN, size,
					0, val);
			}
			else {
				addr -= rsize;
				addloc(name, prim, type, CAUTO, size, addr, 0);
			}
			if (ini && !stat) {
				if (Nli >= MAXLOCINIT) {
					error("too many local initializers",
						NULL);
					Nli = 0;
				}
				LIaddr[Nli] = addr;
				LIval[Nli++] = val;
			}
			if (COMMA == Token)
				Token = scan();
			else
				break;
		}
		semi();
	}
	return addr;
}

static int intcmp(int *x1, int *x2) {
	while (*x1 && *x1 == *x2)
		x1++, x2++;
	return *x1 - *x2;
}

static void signature(int fn, int from, int to) {
	int	types[MAXFNARGS+1], i;

	if (to - from > MAXFNARGS)
		error("too many function parameters", Names[fn]);
	for (i=0; i<MAXFNARGS && from < to; i++)
		types[i] = Prims[--to];
	types[i] = 0;
	if (NULL == Mtext[fn]) {
		Mtext[fn] = galloc((i+1) * sizeof(int), 1);
		memcpy(Mtext[fn], types, (i+1) * sizeof(int));
	}
	else if (intcmp((int *) Mtext[fn], types))
		error("declaration does not match prior prototype: %s",
			Names[fn]);
}

/*
 * decl :=
 *	  declarator { localdecls stmt_list }
 *	| decl_list ;
 *
 * decl_list :=
 *	  declarator
 *	| declarator , decl_list
 */

void decl(int clss, int prim) {
	char	name[NAMELEN+1];
	int	pbase, type, size = 0, val, init;
	int	lsize;

	pbase = prim;
	for (;;) {
		prim = pbase;
		val = 0;
		init = 0;
		type = declarator(0, clss, name, &prim, &size, &val, &init);
		if (TFUNCTION == type) {
			clss = clss == CSTATIC? CSPROTO: CEXTERN;
			Thisfn = addglob(name, prim, type, clss, size, 0,
					NULL, 0);
			signature(Thisfn, Locs, NSYMBOLS);
			if (LBRACE == Token) {
				clss = clss == CSPROTO? CSTATIC:
					clss == CEXTERN? CPUBLIC: clss;
				Thisfn = addglob(name, prim, type, clss, size,
					0, NULL, 0);
				Token = scan();
				lsize = localdecls();
				gentext();
				if (CPUBLIC == clss) genpublic(name);
				genaligntext();
				genname(name);
				genentry();
				genstack(lsize);
				genlocinit();
				Retlab = label();
				compound(0);
				genlab(Retlab);
				genstack(-lsize);
				genexit();
				if (O_debug & D_LSYM)
					dumpsyms("LOCALS: ", name, Locs,
						NSYMBOLS);
			}
			else {
				semi();
			}
			clrlocs();
			return;
		}
		if (CEXTERN == clss && init) {
			error("initialization of 'extern': %s", name);
		}
		addglob(name, prim, type, clss, size, val, NULL, init);
		if (COMMA == Token)
			Token = scan();
		else
			break;
	}
	semi();
}

/*
 * structdecl :=
 *	  STRUCT identifier { member_list } ;
 *
 * member_list :=
 *	  primtype mdecl_list ;
 *	| primtype mdecl_list ; member_list
 *
 * mdecl_list :=
 *	  declarator
 *	| declatator , mdecl_list
 */

void structdecl(int clss, int uniondecl) {
	int	base, prim, size, dummy, type, addr = 0;
	char	name[NAMELEN+1], sname[NAMELEN+1];
	int	y, usize = 0;

	Token = scan();
	copyname(sname, Text);
	ident();
	if (Token != LBRACE) {
		prim = primtype(uniondecl? UNION: STRUCT, sname);
		decl(clss, prim);
		return;
	}
	y = addglob(sname, uniondecl? PUNION: PSTRUCT, TSTRUCT,
			CMEMBER, 0, 0, NULL, 0);
	Token = scan();
	while (	INT == Token || CHAR == Token || VOID == Token ||
		STRUCT == Token || UNION == Token
	) {
		base = primtype(Token, NULL);
		size = 0;
		Token = scan();
		for (;;) {
			if (eofcheck()) return;
			prim = base;
			type = declarator(1, clss, name, &prim, &size,
						&dummy, &dummy);
			addglob(name, prim, type, CMEMBER, size, addr,
				NULL, 0);
			size = objsize(prim, type, size);
			if (size < 0)
				error("size of struct/union member"
					" is unknown: %s",
					name);
			if (uniondecl) {
				usize = size > usize? size: usize;
			}
			else {
				addr += size;
				addr = (addr + INTSIZE-1) / INTSIZE * INTSIZE;
			}
			if (Token != COMMA) break;
			Token = scan();
		}
		semi();
	}
	rbrace();
	semi();
	Sizes[y] = uniondecl? usize: addr;
}

/*
 * top :=
 *	  ENUM enumdecl
 *	| decl
 *	| primtype decl
 *	| storclass decl
 *	| storclass primtype decl
 *
 * storclass :=
 *	  EXTERN
 *	| STATIC
 */

void top(void) {
	int	prim, clss = CPUBLIC;

	switch (Token) {
	case EXTERN:	clss = CEXTERN; Token = scan(); break;
	case STATIC:	clss = CSTATIC; Token = scan(); break;
	case VOLATILE:	Token = scan(); break;
	}
	switch (Token) {
	case ENUM:
		enumdecl(1);
		break;
	case STRUCT:
	case UNION:
		structdecl(clss, UNION == Token);
		break;
	case CHAR:
	case INT:
	case VOID:
		prim = primtype(Token, NULL);
		Token = scan();
		decl(clss, prim);
		break;
	case IDENT:
		decl(clss, PINT);
		break;
	default:
		error("type specifier expected at: %s", Text);
		Token = synch(SEMI);
		break;
	}
}

static void stats(void) {
	printf(	"Memory usage: "
		"Symbols: %5d/%5d, "
		"Names: %5d/%5d, "
		"Nodes: %5d/%5d\n",
		Globs, NSYMBOLS,
		Nbot, POOLSIZE,
		Ndmax, NODEPOOLSZ);
}

void defarg(char *s) {
	char	*p;

	if (NULL == s) return;
	if ((p = strchr(s, '=')) != NULL)
		*p++ = 0;
	else
		p = "";
	addglob(s, 0, TMACRO, 0, 0, 0, globname(p), 0);
	if (*p) *--p = '=';
}

void program(char *name, FILE *in, FILE *out, char *def) {
	init();
	defarg(def);
	Infile = in;
	Outfile = out;
	File = Basefile = name;
	genprelude();
	Token = scan();
	while (XEOF != Token)
		top();
	genpostlude();
	if (O_debug & D_GSYM) dumpsyms("GLOBALS", "", 1, Globs);
	if (O_debug & D_STAT) stats();
}
